azavea/commonhealth-data-map
# White-labeling functionality

## Overview

This PR puts the necessary mechanisms into place for path-based white-labeling. This approach uses the first segment of the url to determine which application configuration to use.

Additionally, this approach introduces feature flags and allows feature flags to be flipped for each application configuration if desired.

The app isn't in TS but I'll describe the config and white-labeling structure here using it
```ts
// each feature flag is an entry in this object with a boolean value

type FeatureFlags {
  [featureName: string]: boolean;
}

type Asset {
  // some types have special handling such as "style"
  // any "style" assets are rendered into `<link>` elements
  // consumers can use "type" to grab the asset they need, such as "logo"
  type: string;
  path: string;
}

type Configuration {
  // if a url is accessed and the code doesn't match the configuration, the default config is used
  code: string;
  default: boolean;
  assets: Asset[];
  strings: {
    [stringName: string]: string
  };
  // each configuration can specify feature-flags
  // which are merged with the default feature flag values
  features: FeatureFlags;
}
```

There must be at least one configuration marked as default, otherwise an error is thrown.

Theming the application is done by creating stylesheets, adding them to the `public` directory, and then adding them to the relevant configuration.

The active config can be accessed using `useContext` like so:

```js
import { ConfigurationContext } from "../ConfigurationProvider";

...

export default () => {
  const config = useContext(ConfigurationContext);
  ...

```

Changing text for each config is done by adding a named string to the `string` portion of the config and then using it 

### Checklist


```console
$ chan added "A feature with no bugs. [#00](https://github.com/azavea/commonhealth-data-map/pull/00)"
```

- [x] Unnecessary `console.log()`s are removed

### Notes

Optional. Ancillary topics, caveats, alternative strategies that didn't work out, anything else.

## Testing Instructions

As it is implemented now (mostly as a demo), there are two configs `dashboard` (the default) and `eac`.

- visit `/eac` and see one thing (notice the table's existence in particular for feature flag usage)
- visit `/` and see another thing

Closes #3
